File: leetcode_75/Sliding_Window/python/1004.py
from typing import List


class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:

        zeros_indexes = list()
        prev_zero = 0
        zeros_counter = 0

        ind = 0
        while ind < len(nums):
            if nums[ind] == 0:
                zeros_indexes.append(ind)
                zeros_counter += 1
                if zeros_counter > k:
                    break
            ind += 1

        mx = ind
        zeros_counter -= 1

        for i in range(ind + 1, len(nums)):
            if nums[i] != 0:
                mx = max(mx, i - zeros_indexes[prev_zero])
            else:
                zeros_indexes.append(i)
                zeros_counter += 1
                mx = max(mx, i - zeros_indexes[prev_zero] - 1)
                prev_zero += 1
                zeros_counter -= 1

        return mx

File: leetcode_75/Sliding_Window/cpp/1004.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int longestOnes(vector<int> &nums, int k)
    {
        vector<int> zeros_indexes;
        int prev_zero = 0;
        int zeros_counter = 0;

        int ind = 0;
        while (ind < nums.size())
        {
            if (nums[ind] == 0)
            {
                zeros_indexes.push_back(ind);
                zeros_counter++;
                if (zeros_counter > k)
                {
                    break;
                }
            }
            ind++;
        }
        int mx = ind;
        zeros_counter--;

        for (int i = ind + 1; i < nums.size(); ++i)
        {
            if (nums[i] != 0)
            {
                mx = max(mx, i - zeros_indexes[prev_zero]);
            }
            else
            {
                zeros_indexes.push_back(i);
                zeros_counter += 1;
                mx = max(mx, i - zeros_indexes[prev_zero] - 1);
                prev_zero++;
                zeros_counter--;
            }
        }

        return mx;
    }
};